raft/src/main/java/org/apache/kafka/raft/ReplicatedCounter.java

```diff
@@ -157,13 +162,27 @@ public synchronized void handleSnapshot(SnapshotReader<Integer> reader) {
     public synchronized void handleLeaderChange(LeaderAndEpoch newLeader) {
         if (newLeader.isLeader(nodeId)) {
             log.debug("Counter uncommitted value initialized to {} after claiming leadership in epoch {}",
-                committed, newLeader);
+                    committed, newLeader);
             uncommitted = committed;
             claimedEpoch = OptionalInt.of(newLeader.epoch());
         } else {
             log.debug("Counter uncommitted value reset after resigning leadership");
             uncommitted = -1;
             claimedEpoch = OptionalInt.empty();
         }
+        handleSnapshotCalled = false;
+        handleSnapshotCalls = 0;
+    }
+
+    public boolean isLeader() {
+        return this.client.leaderAndEpoch().isLeader(nodeId);
```

Comment:
I think this is the issue you reported in the Jira. The `RaftClient.Listener` should not use `RaftClient.leaderAndEpoch` to determine if it is the leader. It should instead use `RaftClient.Listener.handleLeaderChange`.

For this state machine `ReplicatedCounter` we should look at the `claimedEpoch` variable.

I am going to create an issue to remove this method. cc @hachikuji

Reply:
https://issues.apache.org/jira/browse/KAFKA-13006

Reply:
done